### PR TITLE
Reduce flakiness in canary metrics script by checking whether multiple metrics are unhealthy before a rollback

### PR DESCRIPTION
**Pull request by maggie-lou**
**Related issues**: N/A

### CODE REVIEW

File: enterprise/tools/check_metrics/check_metrics.go [hunk 1]
```diff
@@ -95,7 +103,21 @@ func main() {
 					}
 
 					if unhealthyCount >= maxUnhealthyCount {
-						return errors.New(fmt.Sprintf("%s metrics unhealthy.", metric.Name))
+						err = func() error {
```
Comment (iain-macdonald) [inline enterprise/tools/check_metrics/check_metrics.go:106]: why not just do this stuff inline instead of wrapping in a function?

File: enterprise/tools/check_metrics/check_metrics.go [hunk 2]
```diff
@@ -201,3 +223,75 @@ func relativeRangeMetricHealthy(promAPI promapi.API, metricName string, metricVa
...
+
+type TextBlock struct {
+	Type string `json:"type"`
+	Text string `json:"text"`
+}
+
+func sendMetricUnhealthySlack(metric string, isSecondaryMetric bool, maxSecondaryMetricFailureCount int) {
+	slackWebhookURL := os.Getenv("SLACK_WEBHOOK")
+	if slackWebhookURL == "" {
+		log.Warningf("No slack webhook set. Cannot send warning messages.")
```
Comment (iain-macdonald) [inline enterprise/tools/check_metrics/check_metrics.go:245]: Errorf / Fatalf?

File: enterprise/tools/check_metrics/check_metrics.go [hunk 3]
```diff
@@ -95,7 +103,21 @@ func main() {
 					}
 
 					if unhealthyCount >= maxUnhealthyCount {
-						return errors.New(fmt.Sprintf("%s metrics unhealthy.", metric.Name))
+						err = func() error {
+							log.Warningf("Metric %s has been consecutively unhealthy %d times. Marking as failed.", metric.Name, unhealthyCount)
+							sendMetricUnhealthySlack(metric.Name, metric.IsSecondaryMetric, config.MaxSecondaryMetricFailureCount)
```
Comment (iain-macdonald) [inline enterprise/tools/check_metrics/check_metrics.go:108]: WDYT about gathering the status of all metrics and sending them all (including the passing ones) to slack instead of sending them off one-by-one like this?

**Review by iain-macdonald — COMMENTED** [review]
Out of curiosity, would using a moving average for some of the spiky metrics improve their reliability? Curious about that approach in combination with or instead of this approach

**Comment by maggie-lou** [comment]
> Out of curiosity, would using a moving average for some of the spiky metrics improve their reliability? Curious about that approach in combination with or instead of this approach

My understanding is that moving averages are not the best fit for canary scripts. We only want to consider metrics for the canary from after it has been re-deployed. Otherwise we might smooth out changes caused by the deploy. We also don't want the metrics script to run for too long, because if the canary is unhealthy, we want to roll back sooner than later. That means there's only a small window where we're polling the metrics, and at that point if the metrics are spiking, the moving average would spike too

File: enterprise/tools/check_metrics/check_metrics.go (hunk 1 above)
Comment (maggie-lou) [reply enterprise/tools/check_metrics/check_metrics.go]: Fixed

File: enterprise/tools/check_metrics/check_metrics.go (hunk 2 above)
Comment (maggie-lou) [reply enterprise/tools/check_metrics/check_metrics.go]: Changed to Errorf. Didn't want to use Fatalf because this script causes a rollback in the deploy if it fails. If the env var accidentally gets unset and the slack can't be sent, it probably shouldn't trigger a rollback

File: enterprise/tools/check_metrics/check_metrics.go (hunk 3 above)
Comment (maggie-lou) [reply enterprise/tools/check_metrics/check_metrics.go]: Done

File: enterprise/tools/check_metrics/check_metrics.go [hunk 4]
```diff
@@ -201,3 +235,105 @@ func relativeRangeMetricHealthy(promAPI promapi.API, metricName string, metricVa
...
+
+		if m.consecutiveUnhealthyCount > 0 {
+			// Display unhealthy metrics at the top of the message
+			msg = metricOverview + msg
+		} else {
+			msg += metricOverview
+		}
+	}
+
+	msgColor := "#36a64f" // Green
```
Comment (iain-macdonald) [inline enterprise/tools/check_metrics/check_metrics.go:283]: nit: consider defining consts for these colors

File: enterprise/tools/check_metrics/config.go [hunk 5]
```diff
@@ -19,12 +22,17 @@ type PrometheusMetric struct {
...
 	// If the metric does not meet this health threshold, it is considered unhealthy
 	HealthThreshold HealthThreshold `yaml:"health_threshold"`
 	// Set if it's valid for the metric value to be 0 or missing
 	// Ex. For error log count or invocation failure rate, at any given time the value can be 0
 	IsMissingDataValid bool `yaml:"is_missing_data_valid"`
+	// Primary metric failure should immediately result in a rollback
+	// If a secondary metric fails, it should only result in a rollback if several other secondary metrics
+	// are also failing (configurable by MaxSecondaryMetricFailureCount). This is intended to reduce unnecessary
+	// rollbacks from spiky secondary metrics
+	IsSecondaryMetric bool `yaml:"is_secondary_metric"`
```
Comment (iain-macdonald) [inline enterprise/tools/check_metrics/config.go:35]: nit: feel free to shorten this to something like `is_secondary` or something if you want

**Review by iain-macdonald — APPROVED** [review]
